driver/ni/ni.cpp

```diff
@@ -159,111 +577,312 @@ int ni::Source::init(){
-        return freighter::NIL;
+freighter::Error ni::Source::start() {
+    if (this->breaker.running() || !this->ok()) return freighter::NIL;
+    this->breaker.start();
+    if (this->check_ni_error(ni::NiDAQmxInterface::StartTask(this->task_handle))) {
+        this->log_error(
+            "failed while starting reader for task " + this->reader_config.task_name +
+            " requires reconfigure");
+        this->clear_task();
+        return freighter::Error(driver::CRITICAL_HARDWARE_ERROR);
     }
-    this->sample_thread.join();
+    this->sample_thread = std::thread(&ni::Source::acquire_data, this);
+    ctx->setState({
+        .task = task.key,
+        .variant = "success",
+        .details = {
+            {"running", true},
+            {"message", "Task started successfully"}
+        }
+    });
+    return freighter::NIL;
+}
 
-    if (this->checkNIError(ni::NiDAQmxInterface::StopTask(this->task_handle))){
-        LOG(ERROR) << "[NI Reader] failed while stopping reader for task " << this->reader_config.task_name;
+freighter::Error ni::Source::stop() {
+    if (!this->breaker.running() || !this->ok()) return freighter::NIL;
+    this->breaker.stop();
+    if (this->sample_thread.joinable()) this->sample_thread.join();
+    if (this->check_ni_error(ni::NiDAQmxInterface::StopTask(this->task_handle))) {
+        this->log_error(
+            "failed while stopping reader for task " + this->reader_config.task_name);
         return freighter::Error(driver::CRITICAL_HARDWARE_ERROR);
     }
-    LOG(INFO) << "[NI Reader] successfully stopped  reader for task " << this->reader_config.task_name;
     data_queue.reset();
-    return  freighter::NIL;
-
+    LOG(INFO) << "[NI Reader] stopped reader for task " << this->reader_config.
+            task_name;
+    ctx->setState({
+        .task = task.key,
+        .variant = "success",
+        .details = {
+            {"running", false},
+            {"message", "Task stopped successfully"}
+        }
+    });
+    return freighter::NIL;
 }
 
 
-void ni::Source::clearTask(){
-    if (this->checkNIError(ni::NiDAQmxInterface::ClearTask(this->task_handle))){
-        LOG(ERROR) << "[NI Reader] failed while clearing reader for task " << this->reader_config.task_name;
+void ni::Source::clear_task() {
+    if (this->check_ni_error(ni::NiDAQmxInterface::ClearTask(this->task_handle))) {
+        this->log_error(
+            "failed while clearing reader for task " + this->reader_config.task_name);
     }
-    LOG(INFO) << "[NI Reader] cleared reader for task " << this->reader_config.task_name;
 }
 
 
-ni::Source::~Source(){
-   this->clearTask();
+ni::Source::~Source() {
+    this->clear_task();
+    if(this->sample_thread.joinable()) this->sample_thread.join();
+    LOG(INFO) << "[NI Reader] joined sample thread";
 }
 
-int ni::Source::checkNIError(int32 error){
-    if (error < 0){
-        char errBuff[2048] = {'\0'};
+int ni::Source::check_ni_error(int32 error) {
+    if (error < 0) {
```

Comment:
Reversing the logic would make it cleaner here. I.e. do `if (error >= 0) return`
and then you don't need brackets inside if statement